### PR TITLE
Data API Template App

### PR DESCRIPTION
**Pull request by nlarew**
### Changes:

- Adds a new backend for the Data API template
- Adds a new `prod-data-api` state to the `web-js` frontend

~**NOTE: This is waiting on upstream changes to how we generate `realm.json` - we need deployment info and/or API base URLs for the app. See https://jira.mongodb.org/browse/BAAS-19713**~

### To test:

1. Navigate to the project root
   
   ```
   cd other/web-js
   ```

2. Create a backend app
   
   ```
   cd backend-data-api
   npx mongodb-realm-cli push
   # Note your new App ID
   cd ..
   ```

3. Update `client/src/realm.json`
   
   ```
   {
     "appId": "<CLIENT APP ID>",
     "baseUrl": "https://realm.mongodb.com",
     "appUrl": "https://realm.mongodb.com/groups/<ATLAS PROJECT ID>/apps/<ATLAS APP ID>",
     "dataSourceName": "mongodb-atlas"
   }
   ```
You can get `appId` and `appUrl` from the UI:
- `appId` is the one with your App name, e.g. `myapp-abcde`
- `appUrl` you can copy from the browser URL bar

4. Run the combined tests - these run on the development code for the current state (set in the `.env` file)
   
   ```
   cd client
   npm install
   npm run test:mql
   npm run test:graphql
   npm run test:data-api
   cd ..
   ```

5. Run the artifact tests - these run on the generated code for each state. Do this for each artifact repo.
   
   ```
   # First, copy realm.json into the artifact projects
   xargs -n 1 cp client/src/realm.json<<<"generated/prod-graphql/src/ generated/prod-mql/src/ generated/prod-data-api/src/"
   # Then, install and test each artifact project
   cd generated/prod-data-api
   # cd generated/prod-mql
   # cd generated/prod-graphql
   npm install
   npm run test
   cd ../..
   ```

### CODE REVIEW

File: other/web-js/bluehawk.sh [hunk 1]
```diff
@@ -2,10 +2,135 @@
...
+      jq_expr="${jq_expr}, ${dep_str}"
+    fi
+  done
+  jq_expr="del(${jq_expr})"
+
+  jq "$jq_expr" package.json >> package.json.tmp
+  mv package.json.tmp package.json
+}
+
+function modify_craco_config() {
```
Comment (mongodben) [inline other/web-js/bluehawk.sh:45]: we should migrate these templates from CRA to another web framework sometime soonish b/c CRA is no longer actively maintained...not for this PR tho

File: other/web-js/client/src/client-api.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    if (deployment_model === "GLOBAL") {
+      return `https://realm.mongodb.com/api/client/v2.0/`;
+    }
+    if (!cloud || !region) {
+      throw new Error(
+        `Must specify a cloud provider and region for LOCAL Apps. e.g. { "cloud": "aws", "region": "us-east-1" }`
+      );
+    }
+    return `https://${region}.${cloud}.realm.mongodb.com/api/client/v2.0/`;
+  }
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:18]: why are we constructing the URL rather than just passing the string?

File: other/web-js/client/src/client-api.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+      deployment_model:
+        deployment_model ?? (cloud || region) ? "LOCAL" : "GLOBAL",
+      cloud,
+      region,
+    };
+    this.baseUrl = ClientApi.constructBaseUrl(this.deployment);
+    this.onAuthChange = onAuthChange;
+  }
+
+  // endpointUrl(path: string): string;
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:34]: rather than these comments showing what's going on, can you refactor them using JSDoc?

here for example, it'd be:

```suggestion
  /**
  * @param {string} path
  * @returns {string}
  */
```

File: other/web-js/client/src/client-api.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      cache: "no-cache",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(credentials),
+    });
+    if (resp.status === 200) {
+      // Example 200 Response:
+      // {
+      //   access_token: "eyJhbGciOiJIUzI1NiIsInA5cCI6IkpXVCJ9.eyJiYWFzX2RldmljZV9pZCI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5YSIsImJhYXNfZG9tYWluX2lkIjoiNWNkYjEyNDA4ZTIzMmFjNGY5NTg3ZmU4IiwiZXhwIjoxNjc3MDAzNjk1LCJpYXQiOjE2NzcwMDE4OTUsImlzcyI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5ZCIsInN0aXRjaF9kZXZJZCI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5YSIsInN0aXRjaF9kb21haW5JZCI6IjVjZGIxMjQwOGUyMzJhYzRmOTU4N2ZlOCIsInN1YiI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5NSIsInR5cCI6ImFjY2VzcyJ9.R9ZazXGPtngBfaMe1nZkKPjvoPRe3GWNfA77UX28ewo",
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:85]: these are real chunky. maybe do some `...` in the middle of the token strings?

File: other/web-js/client/src/client-api.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      return;
+    } else {
+      throw new ClientApiError(await resp.json());
+    }
+  };
+
+  // logIn(provider: string, credentials: any): Promise<void>
+  logIn = async (provider, credentials) => {
+    const url = this.endpointUrl(`/auth/providers/${provider}/login`);
+    const resp = await fetch(url, {
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:74]: nit: consider renaming this variable to `response`, as that's what's in fetch API docs https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

and then the result of `await response.json()` to `data`

File: other/web-js/client/src/client-api.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+      //   device_id: "63f504a75dd881f2f58ba09a",
+      //   refresh_token: "eyJhbGciOiJIUzI1BiIsInE5cCI6IkpXVCJ9.eyJiYWFzX2RhdGEiOm51bGwsImJhYXNfZGV2aWNlX2lkIjoiNjNmNTA0YTc1ZGRlOTFlNmY2MGJhMDlhIiwiYmFhc19kb21haW5faWQiOiI1Y2RiMTI0MDhlMjMyYWM0Zjk1ODdmZTgiLCJiYWFzX2lkIjoiNjNmNTA0YTc1ZGRlOTFlNmY2MGJhMDlkIiwiYmFhc19pZGVudGl0eSI6eyJpZCI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5Mi1qenBncGhsY3puZXpxaXJkb3ljemNwc20iLCJwcm92aWRlcl90eXBlIjoiYW5vbi11c2VyIiwicHJvdmlkZXJfaWQiOiI1ZDQ4OTJlMzU1OGViYjk4Y2VkNmU5MWYifSwiZXhwIjozMjUzODAxODk1LCJpYXQiOjE2NzcwMDE4OTUsInN0aXRjaF9kYXRhIjpudWxsLCJzdGl0Y2hfZGV2SWQiOiI2M2Y1MDRhNzVkZGU5MWU2ZjYwYmEwOWEiLCJzdGl0Y2hfZG9tYWluSWQiOiI1Y2RiMTI0MDhlMjMyYWM0Zjk1ODdmZTgiLCJzdGl0Y2hfaWQiOiI2M2Y1MDRhNzVkZGU5MWU2ZjYwYmEwOWQiLCJzdGl0Y2hfaWRlbnQiOnsiaWQiOiI2M2Y1MDRhNzVkZGU5MWU2ZjYwYmEwOTItanpwZ3BobGN6bmV6cWlyZG95Y3pjcHNtIiwicHJvdmlkZXJfdHlwZSI6ImFub24tdXNlciIsInByb3ZpZGVyX2lkIjoiNWQ0ODkyZTM1NThlYmI5OGNlZDZlOTFmIn0sInN1YiI6IjYzZjUwNGE3NWRkZTkxZTZmNjBiYTA5NSIsInR5cCI6InJlZnJlc2gifQ.k1o_4RQ1diSewmXynIqiC0bxVAdvYJmw3L1358T3UJk",
+      //   user_id: "63f504a25dde21e6f10ba025"
+      // }
+      const response = await resp.json();
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:90]: following above comment https://github.com/mongodb-university/realm-template-apps/pull/142/files#r1135580647, consider renaming this variable to `data`

File: other/web-js/client/src/client-api.js [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    if (resp.status === 201) {
+      // Status 201 means the user was created. There is no response body.
+      return;
+    } else {
+      throw new ClientApiError(await resp.json());
+    }
+  };
+
+  // logIn(provider: string, credentials: any): Promise<void>
+  logIn = async (provider, credentials) => {
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:72]: higher level question that i just noticed upon reviewing this method: 

why are we using the HTTP auth here instead of the Web SDK auth client which'd abstract all this stuff away from us having to implement?

File: other/web-js/client/src/client-api.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+ * @property {string} error_code - A machine-readable error code.
+ * @property {string} link - A link to a related application log in the App Services UI.
+ */
+export class ClientApiError extends Error {
+  // constructor(input: { error: string, error_code: string, link: string }): ClientApiError
+  constructor({ error, error_code, link }) {
+    super(error);
+    this.name = "ClientApiError";
+    this.error = error;
+    this.error_code = error_code;
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:209]:
```suggestion
    this.errorCode = error_code;
```

nit: should this be camelCase like your other user-defined variables?

File: other/web-js/client/src/components/App.test.js [hunk 9]
```diff
@@ -1,8 +1,139 @@
-import { render, screen } from "@testing-library/react";
+import { render, waitFor, screen, cleanup } from "@testing-library/react";
```
Comment (mongodben) [inline other/web-js/client/src/components/App.test.js:1]: nice inclusion of these tests 👍

File: other/web-js/client/src/components/AppName.js [hunk 10]
```diff
@@ -6,6 +6,7 @@ const API_TYPE = process.env.REACT_APP_API_TYPE;
 const API_TYPE_NAMES = {
   graphql: "GraphQL",
   mql: "MQL",
+  "data-api": "Data API",
   local: "Local",
 };
```
Comment (mongodben) [inline other/web-js/client/src/components/AppName.js:11]: idea: rather than having this here + additional metadata in `other/web-js/client/src/components/MoreInfo.js`, maybe create a `metadata.js` file with API type names, docs link, and any other API type-specific config

File: other/web-js/client/src/components/Theme.js [hunk 11]
```diff
@@ -27,7 +27,15 @@ switch (API_TYPE) {
     };
     break;
   }
-  // :state-start: development
+  case "data-api": {
```
Comment (mongodben) [inline other/web-js/client/src/components/Theme.js:30]: this info could also be in the metadata.js file

File: other/web-js/client/src/components/TodoItem.js [hunk 12]
```diff
@@ -14,6 +14,7 @@ export function TodoItem({ todo, todoActions }) {
     <ListItem>
       <ListItemIcon>
         <Checkbox
+          data-testid="todo-checkbox"
```
Comment (mongodben) [inline other/web-js/client/src/components/TodoItem.js:17]: should this be bluehawk removed?

File: other/web-js/client/src/components/Theme.js [hunk 13]
```diff
@@ -37,7 +45,6 @@ switch (API_TYPE) {
     };
     break;
   }
-  // :state-end:
   default: {
     throw new Error(
       `Invalid REACT_APP_API_TYPE: "${API_TYPE}". Specify "graphql", "mql", or "local" instead.`
```
Comment (mongodben) [inline other/web-js/client/src/components/Theme.js:50]: > Specify "graphql", "mql", or "local" instead.

these API types probably should come from app metadata rather than be hardcoded here

File: other/web-js/client/src/client-api.js (hunk 2 above)
Comment (nlarew) [reply other/web-js/client/src/client-api.js:18]: Don't have access to the strings in realm.json yet. Will use them directly when we can!

File: other/web-js/client/src/components/WelcomePage.js [hunk 14]
```diff
@@ -77,6 +122,7 @@ export function WelcomePage() {
           <NonAuthErrorAlert />
           <TextField
             id="input-email"
+            data-testid="input-email"
```
Comment (mongodben) [inline other/web-js/client/src/components/WelcomePage.js:125]: should these `data-testid` props be removed? applies to all others as well

File: other/web-js/client/src/client-api.js (hunk 7 above)
Comment (nlarew) [reply other/web-js/client/src/client-api.js:72]: Frankly a big takeaway for me is that you probably should just use the web SDK. That said this shows how to use just HTTP to work with the Data API from a web app. Pulling in the web SDK only to use the data api instead feels strange.

File: other/web-js/client/src/data-api.js [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+ */
+export class DataApi {
+  static constructBaseUrl(appId, location) {
+    const { deployment_model, cloud, region } = location;
+    if (deployment_model === "LOCAL") {
+      return `https://${region}.${cloud}.data.mongodb-api.com/app/${appId}/endpoint/data/v1/`;
+    } else {
+      return `https://data.mongodb-api.com/app/${appId}/endpoint/data/v1/`;
+    }
+  }
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:14]: this seems somewhat duplicative of the logic in https://github.com/mongodb-university/realm-template-apps/pull/142/files#diff-3e277fd749d5f7165b2ef838367404e1143d07bc50d05ef64a0b63f2676f33eeR7-R18.

is there a way to consolidate these?

File: other/web-js/client/src/data-api.js [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+
+  // constructor(config: {
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:16]: can you convert this and all method annotations to JSDoc?

File: other/web-js/client/src/client-api.js (hunk 8 above)
Comment (nlarew) [reply other/web-js/client/src/client-api.js:209]: I think I went with snake here to most closely match the underlying error. Doesn't matter to me though, can update.

File: other/web-js/client/src/components/TodoItem.js (hunk 12 above)
Comment (nlarew) [reply other/web-js/client/src/components/TodoItem.js:17]: Nope the tests work on the generated artifact projects too

File: other/web-js/client/src/data-api.js [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+    const response = await resp.json();
+
+    if (resp.status === 200 || resp.status === 201) {
+      return response;
+    } else {
+      throw new Error(response);
+    }
+  }
+
+  // type FindOneInput = {
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:100]: note that you can use JSDoc [@typedef](https://jsdoc.app/tags-typedef.html) here, which'd provide intellisense. though i think the TS types in comments are easier to read if you're actually looking at the source code

File: other/web-js/client/src/hooks/useDataApi.js [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+      findOne: api.findOne,
+      find: api.find,
+      insertOne: api.insertOne,
+      insertMany: api.insertMany,
+      updateOne: api.updateOne,
+      updateMany: api.updateMany,
+      replaceOne: api.replaceOne,
+      deleteOne: api.deleteOne,
+      deleteMany: api.deleteMany,
+      // aggregate: api.aggregate,
```
Comment (mongodben) [inline other/web-js/client/src/hooks/useDataApi.js:44]: why commented out?

File: other/web-js/client/src/hooks/useDataApi.js [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }, [api, currentUser]);
```
Comment (mongodben) [inline other/web-js/client/src/hooks/useDataApi.js:46]: for my own react education, why are you memoizing each of these individually instead of memoizing just the `api` obj?

Comment (nlarew) [reply other/web-js/client/src/hooks/useDataApi.js:46]: Needed to keep currentUser in state which mucks up returning just the class

Comment (nlarew) [reply other/web-js/client/src/hooks/useDataApi.js:46]: And I could memoize just the API class instance but I'd need to nest it under another field in the return object like `{ currentUser: ..., api: ... }`

File: other/web-js/client/src/hooks/useCollection.js [hunk 20]
```diff
@@ -1,5 +1,5 @@
 import React from "react";
-import { useRealmApp } from "../components/RealmApp";
+import { useApp } from "../components/RealmApp";
```
[the remark this answers is not in the document]
Comment (nlarew) [reply other/web-js/client/src/hooks/useCollection.js:2]: Oh yeah forgot I did that but yes it'll only show up on the states that use an SDK

File: other/web-js/client/src/hooks/useTodos_data-api.js [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+        setTodos((oldTodos) => {
+          const idx = oldTodos.length;
+          return addValueAtIndex(oldTodos, idx, {
+            ...document,
+          });
+        });
+      } catch (err) {
+        if (err.error?.match(/^Duplicate key error/)) {
+          console.warn(
+            `The following error means that we tried to insert a todo multiple times (i.e. an existing todo has the same _id). In this app we just catch the error and move on. In your app, you might want to debounce the save input or implement an additional loading state to avoid sending the request in the first place.`
```
Comment (mongodben) [inline other/web-js/client/src/hooks/useTodos_data-api.js:64]: rather than 'we', consider refactoring to be 'this app' to be more specific

**Review by mongodben — COMMENTED** [review]
some smaller comments throughout and also this one about trouble i had running the code https://github.com/mongodb-university/realm-template-apps/pull/142#issuecomment-1468353714

also, i think it's outside the scope of this PR, but it's worth re-considering the structure of the Web template apps for the following reasons:

1. the way that it currently stands with all the bluehawk states is quite hard to follow. also there's no way to test the code at the source for 2 of the 3 ways w/o bluehawking. i wonder if there's a way to refactor the project to use less bluehawk state. for example, abstracting the MongoDB-interaction interface into a since class that calls one of the three services (MQL, data api, GQL api) depending on some config/env variable. 
2. move away from CRA/CRACO). while CRA isn't officially deprecated, [it is basically unmaintained](https://github.com/facebook/create-react-app/issues/11180), and there are some funky bugs in it. maybe we could do next.js or vite?

File: other/web-js/client/src/components/WelcomePage.js (hunk 14 above)
Comment (nlarew) [reply other/web-js/client/src/components/WelcomePage.js:125]: Actually can remove this one but only because it's unused in the tests

File: other/web-js/client/src/hooks/useDataApi.js (hunk 18 above)
Comment (nlarew) [reply other/web-js/client/src/hooks/useDataApi.js:44]: Just never got around to implementing it - added and uncommented!

**Comment by takameyer** [comment]
Why not TypeScript 😅

File: other/web-js/client/src/client-api.js [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+    this.baseUrl = ClientApi.constructBaseUrl(this.deployment);
+    this.onAuthChange = onAuthChange;
+  }
+
+  /**
+   * Construct a Client API URL at the given path
+   * @param {string} path The path to the Client API endpoint, e.g. `/auth/providers/local-userpass/login`.
+   * @returns {string} The full URL of the Client API endpoint.
+   */
+  #endpointUrl = (path) => {
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:58]: TIL JS got private methods

File: other/web-js/client/src/client-api.js [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+
+  /**
+   * Log the current user out.
+   * @returns {Promise<void>}
+   * @example
+   * await clientApi.logOut();
+   */
+  logOut = async () => {
+    this.#setCurrentUser(null);
+  };
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:165]: i don't think this needs to be an async fxn

File: other/web-js/client/src/client-api.js [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+   * @param {User | null} user The user to set as the current user, or null to unauthenticate the client.
+   * @example
+   * this.#setCurrentUser({
+   *   "user_id": "63f504a25dde21e6f10ba025"
+   *   "access_token": "eyJhbGciOiJIUzI1NiIsI...TYNZYyCM3RKl77FuW_kBf29POr24"
+   *   "device_id": "63f504a75dd881f2f58ba09a",
+   *   "refresh_token": "eyJhbGciOiJIUzI1BiIsI...ynIqiC0bxVAdvYJmw3L1358T3UJk",
+   * })
+   */
+  #setCurrentUser = (user) => {
```
Comment (mongodben) [inline other/web-js/client/src/client-api.js:85]: this comment applies to all the methods on this class. i notice that your methods are all arrow functions (as opposed to `function(){}` declarations. and also these methods make frequent use of the `this` keyword. 

arrow functions can have a funky relationship with `this` when used in JS classes (see [this article](https://www.codementor.io/@dariogarciamoya/understanding-this-in-javascript-with-arrow-functions-gcpjwfyuc#:~:text=In%20my%20opinion%2C%20arrow%20functions%20are%20the%20best%20choice%20when%20working%20with%20closures%20or%20callbacks%2C%20but%20not%20a%20good%20choice%20when%20working%20with%20class/object%20methods%20or%20constructors.)). 

i haven't looked thru your code to make sure that none of this 'funk' is creeping in. but in order to mitigate against this happening either in the template app or a user adopting the template app consider replacing all arrow function methods with 'normal' function declaration methods.

File: other/web-js/client/src/data-api.js [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+   * @param {string} config.appId The Client App ID of the App Services App to connect to, e.g. `myapp-abcde`.
+   * @param {string} [config.cloud] The cloud provider to connect to. Required if `region` is specified.
+   * @param {string} [config.region] The region to connect to. Required if `cloud` is specified.
+   * @param {function} [config.onAuthChange] A callback that's run with the latest auth state whenever the current user changes.
+   * @example
+   * const dataApi = new DataApi({
+   *   appId: "myapp-abcde",
+   *   cloud: "aws",
+   *   region: "us-east-1",
+   *    onAuthChange: (currentUser) => {
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:28]:
```suggestion
   *   onAuthChange: (currentUser) => {
```

File: other/web-js/client/src/data-api.js [hunk 26]
```diff
@@ -0,0 +1,358 @@
...
+    this.currentUser = this.client.currentUser;
+  }
+
+  /**
+   * Register a new user with the specified authentication provider.
+   * @param {string} provider The name of the authentication provider to use.
+   * @param {object} credentials Information used to authenticate with the specified provider.
+   * @returns {Promise<void>}
+   */
+  registerUser = async (provider, credentials) => {
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:54]: same comment as above about arrow function methods https://github.com/mongodb-university/realm-template-apps/pull/142/files#r1142163413

File: other/web-js/client/src/utils.js [hunk 27]
```diff
@@ -32,10 +47,54 @@ export function removeValueAtIndex(arr, idx) {
   return [...arr.slice(0, idx), ...arr.slice(idx + 1)];
 }
 
+export const createObjectId = () => {
+  // :state-start: development
+  return API_TYPE_NAME === "Data API"
+    ? EJSON.serialize(new ObjectId())
+    : new Realm.BSON.ObjectId();
```
Comment (mongodben) [inline other/web-js/client/src/utils.js:54]:
```suggestion
    ? EJSON.serialize(new ObjectId())
    : new ObjectId();
```

File: other/web-js/client/src/client-api.js (hunk 24 above)
Comment (nlarew) [reply other/web-js/client/src/client-api.js:85]: The binding inference is the point here - otherwise you need to call a bunch of `.bind(this)` in the hooks

File: other/web-js/client/src/utils.js (hunk 27 above)
Comment (nlarew) [reply other/web-js/client/src/utils.js:54]: Gonna leave this as-is to be as close to the generated versions as possible since Realm's bundled BSON package isn't guaranteed to be the full `bson` module.

File: other/web-js/client/craco.config.js [hunk 28]
```diff
@@ -0,0 +1,16 @@
+module.exports = {
```
Comment (mongodben) [inline other/web-js/client/craco.config.js:1]: can we delete this if now using vite?

File: other/web-js/client/craco.config.js
Comment (mongodben) [inline other/web-js/client/craco.config.js:1]: can we delete this file if now using vite?

File: other/web-js/client/package.json [hunk 29]
```diff
@@ -11,22 +11,29 @@
     "@testing-library/jest-dom": "^5.11.4",
     "@testing-library/react": "^13.4.0",
     "@testing-library/user-event": "^14.4.3",
-    "dedent": "^0.7.0",
+    "bson": "^5.0.1",
+    "buffer": "^6.0.3",
+    "crypto-browserify": "^3.12.0",
```
Comment (mongodben) [inline other/web-js/client/package.json:16]: can this (and maybe other dependencies) be deleted since moving to vite? 

stream-browserify and buffer perhaps?

Comment (mongodben) [reply other/web-js/client/package.json:16]: also the craco build/test commands and dev dependency for craco

File: other/web-js/client/src/data-api.js [hunk 30]
```diff
@@ -0,0 +1,358 @@
...
+   */
+
+  /**
+   * Find a single document in a collection.
+   * @param {FindOneInput} input - The request body for the action.
+   * @returns {Promise<FindOneResult>} - The response body for the action.
+   *
+   * @see https://mongodb.com/docs/atlas/app-services/data-api/generated-endpoints/#std-label-data-api-findOne
+   */
+  findOne = async (input) => {
```
Comment (mongodben) [inline other/web-js/client/src/data-api.js:177]: you could consolidate the syntax by defining as:

```suggestion
async findOne(input) {
```
